### ISSUE TITLE
Sync two repositories?

### ISSUE THREAD

**Issue by marbetschar** [LucianoPolit/RepositoryKit#2]
How do you think is the best way/strategy to sync two different repositories, containing the same type of data?

I'd love to use a event based system if the repository supports it (see https://github.com/LucianoPolit/RepositoryKit/issues/1).

**Comment by LucianoPolit** [comment]
Hi again, nice question!

The best way of doing this is having three repositories. Let's use an example:

- The repository `A` interacts with the local storage.
- The repository `B` interacts with a RESTful API.
- The repository `C` is the container of both repositories. It's the one that knows the politics of how to interact with both sub repositories, `A` and `B`. [This](https://github.com/LucianoPolit/RepositoryKit/blob/master/Source/CRUD/CRUDRepository%2BNetworkingStorage.swift) includes a default way to CRUD between two repositories. If, for any reason, you need to do any operation by another way, you can implement your own methods.

So, by this way, you know that everything that is on `A`, is on `B`. But now there are two more problems:

- `B` is not reachable because internet problems.
- `B` can get new data from another place (like your web app or whatever that can creates new entities on the API) and neither `A` or `C` will know.

So, first, lets say that there is a new method that is able to synchronize both repositories. From `A` to `B` and from `B` to `A`. If you really need to synchronize them at any moment, you can call it when internet comes back (not a common case). Anyway, this method should be called the first time the data is required.

Now, with this new method, you can trust that the local storage has the latest data when you need it.

What this method does? You can follow [this](https://github.com/LucianoPolit/RepositoryKit/blob/master/Source/Sync/SynchronizableRepository%2BCRUDNetworkingStorage.swift) for full details.

In summary:
- You need to add a `synchronized` parameter and set it to false before making an operation that requires synchronization (like patching the entity).
- You need a new endpoint `path/collection (POST)` that receives all the entities that have been changed while internet was not reachable.

If you want you can see a full [example](https://github.com/LucianoPolit/RepositoryKit/tree/master/Example) of how to do it. I recommend you to focus on the [entity](https://github.com/LucianoPolit/RepositoryKit/blob/master/Example/Source/User.swift) and the [view controller](https://github.com/LucianoPolit/RepositoryKit/blob/master/Example/Source/UserTableViewController.swift)

Let me know if you would like to know something else!

**Comment by marbetschar** [comment]
I think I understand the basic concept. What I'm not quite sure is, when and how to call the synchronization:

For the local repo we may use the database change notification (preferred) or a regular interval. For remote repos (REST API) we often usually do not have any change notification - so a regular interval seems to be the only option here.

How do you run code on a regular interval reliaby on iOS? As far as I've understood every background task will be suspended, after the app was inactive for about 10mins. How to circumvent this? Do you schedule a local push notification to kick things off? Are there any limits we have to keep an eye on?

Or is there any other way I miss here to make sure we always have the data up2date on the device as soon as the user opens the app?

**Comment by LucianoPolit** [comment]
The base of my reasoning was to trust on the data that you have locally.

So, you should synchronize the repositories when:
- The view controller is presented.
- The app becomes active.
- The app starts (if really needed from the beginning).
- An interval (if really needed to have the last at any moment) -> be careful with that one.

The moment when you call the synchronize method depends entirely on your requirements. And about the app life cycle, you can listen on the `AppDelegate` or observing the `NotificationCenter`. I would recommend you to google a little about it.
